### PR TITLE
Adds IT that verifies scans see data written by concurrent writers

### PR DESCRIPTION
**Pull request by keith-turner**
The IT ensures that when there are concurrent writes, scans, and table operations that scans always see any data written before the scan started.

### CODE REVIEW

**Comment by keith-turner** [comment]
Created this IT for testing #3634, however realized its good for 2.1.X and #3573 to stress read patterns the metadata table relies on, so decided to create it in 2.1 and merge it forward to elasticity.

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+          int splitsToAdd = random.nextInt(10);
+          TreeSet<Text> splits = new TreeSet<>();
+
+          for (int i = 0; i < splitsToAdd; i++) {
+            splits.add(new Text(String.format("%016x", nextLongAbs(random))));
+          }
+
+          tctx.client.tableOperations().addSplits(tctx.table, splits);
+          numSplits += splitsToAdd;
+        }
```
Comment (dlmarion) [inline test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:473]: should we add merging?

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+      }
+
+      for (int i = 0; i < numsScanTask; i++) {
+        scanTasks.add(executor.submit(new ScanTask(testContext)));
+      }
+
+      var tableOpsTask = executor.submit(new TableOpsTask(testContext));
+
+      // let the concurrent mayhem run for a bit
+      Thread.sleep(60000);
```
Comment (dlmarion) [inline test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:100]: Is there any value to letting this run longer?

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+          int rowsToGenerate = random.nextInt(1000);
+
+          try (var writer = tctx.client.createBatchWriter(tctx.table)) {
+            for (int i = 0; i < rowsToGenerate; i++) {
+              Mutation m = generateMutation(random);
+              writer.addMutation(m);
+              stats.written += m.getUpdates().size();
+              dataAdded.add(m);
+            }
```
Comment (dlmarion) [inline test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:401]: Is it possible that the BatchWriter flushes in this loop based on default configs? If so, a scan that starts during this loop will see more data than expected.

**Comment by dlmarion** [comment]
This test is a great idea

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java (hunk 3 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:401]: Yeah that is ok, the scans do not care if they see extra data. They just require that they see all data that was successfully written before the scan started.

There is a very last scan done after all the writers have stopped, that scan does not tolerate extra data and should throw an exception if it sees it.

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java (hunk 2 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:100]: Yeah I think there is value in running it longer, but maybe not in the IT.  We definitely want to know if does not work on a short run.  I think I will add a main method  that allows running it.  I think it would be good to have as an IT and something that is easy to run for longer periods.

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java (hunk 1 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:473]: yeah that would be good to add.  maybe bulk import too

**Comment by keith-turner** [comment]
I added merge, bulk import, compaction filtering, and batch scan to the test in 7cbc635.  I also added some debugging tips based on a problem I ran into where bulk importing deletes was causing the test to fail.  Spent hours trying to track down the cause and finally developed a technique that made tracking the problem down really quick.  I documented the technique [in the code](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L93-L108) and left a [utility method](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L668-L696) related to the technique in the test.

The problem I ran into is documented [here in the test](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L488-L496).  The problem is caused by the fact that major compactions no longer force a flush, this is a change made some time in the past, trying to find an issue for that.  Not flushing before a full major compaction is ok when only doing live ingest or only doing bulk import, but for a mix of bulk import and live ingest not flushing before full major compaction has a consequence as I discovered in the test.

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+      this.tctx = testContext;
+    }
+
+    private long bulkImport(Random random, Collection<Mutation> mutations) throws Exception {
+
+      if (mutations.isEmpty()) {
+        return 0;
+      }
+
+      Path bulkDir = new Path(tctx.tmpDir + "/" + "bulkimport_" + nextLongAbs(random));
```
Comment (EdColeman) [inline test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:437]:
```suggestion
      Path bulkDir = new Path(tctx.tmpDir + "/bulkimport_" + nextLongAbs(random));
```

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java (hunk 2 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:136]: Opened #3646

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java (hunk 1 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java]: added merge and other stuff in 7cbc635

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java [hunk 5]
```diff
@@ -105,7 +105,7 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return size + " " + numEntries;
+    return size + " " + numEntries + " " + time;
```
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java:108]: I realize that this adds to the existing pattern - but would it be worth it to use a more standardized toString format that includes the field names? Otherwise, you need to read the code to understand that '12 1000 1234` is 'size: 12, numEntries 1000, time 1234'

**Review by EdColeman — APPROVED** [review]
Minor suggestion on the toString, but that can be ignored if there is a reason not to use a standard toString format,

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java (hunk 5 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java:108]: It would be good to improve the toString.  I am thinking maybe I should not make this change in 2.1 because I didn't analyze how this toString method was used. I made the change when trying to figure out the problem with bulk importing delete keys.  I'll either improve it or not change it after looking into how its used.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java]: That class is used in a huge number of places.  I decided to err on the side of caution and not change its toString rather try to analyze everywhere its used.

File: test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+import com.google.common.collect.Collections2;
+import com.google.common.collect.MoreCollectors;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * This test verifies that scans will always see data written before the scan started even when
+ * there are concurrent scans, writes, and table operations running.
+ */
+@Tag(SUNNY_DAY)
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:82]: I don't think it makes sense to add this to the sunny tests, as it's not testing base functionality, but it's testing more complicated edge cases that can be left in the larger test suite.
